File: python/x402/mechanisms/evm/batch_settlement/encoding.py
"""ABI-encoding helpers for the batch-settlement deposit collectors."""

from __future__ import annotations

try:
    from eth_abi import encode as abi_encode
    from eth_utils import keccak
except ImportError as e:
    raise ImportError(
        "EVM mechanism requires ethereum packages. Install with: pip install x402[evm]"
    ) from e


def _to_int(hex_or_int: str | int) -> int:
    if isinstance(hex_or_int, int):
        return hex_or_int
    s = hex_or_int.strip()
    if s.startswith("0x") or s.startswith("0X"):
        return int(s, 16)
    return int(s)


def _to_bytes(value: str | bytes) -> bytes:
    if isinstance(value, bytes):
        return value
    return bytes.fromhex(value.removeprefix("0x"))


def build_erc3009_deposit_nonce(channel_id: str, salt: str) -> str:
    """Compute the ERC-3009 nonce used by the deposit collector.

    `keccak256(abi.encode(bytes32 channelId, uint256 salt))`

    Args:
        channel_id: 0x-prefixed bytes32 channel id.
        salt: Random per-deposit salt (hex string of 32 bytes).

    Returns:
        0x-prefixed bytes32 nonce.
    """
    encoded = abi_encode(["bytes32", "uint256"], [_to_bytes(channel_id), _to_int(salt)])
    return "0x" + keccak(encoded).hex()


def build_erc3009_collector_data(
    valid_after: str,
    valid_before: str,
    salt: str,
    signature: str,
) -> bytes:
    """Encode `collectorData` for `ERC3009DepositCollector.collect()`.

    `abi.encode(uint256 validAfter, uint256 validBefore, uint256 salt, bytes signature)`
    """
    return abi_encode(
        ["uint256", "uint256", "uint256", "bytes"],
        [
            _to_int(valid_after),
            _to_int(valid_before),
            _to_int(salt),
            _to_bytes(signature),
        ],
    )


def build_eip2612_permit_data(
    value: str,
    deadline: str,
    v: int,
    r: str,
    s: str,
) -> bytes:
    """Encode an optional EIP-2612 permit segment for `Permit2DepositCollector`.

    `abi.encode(uint256 value, uint256 deadline, uint8 v, bytes32 r, bytes32 s)`
    """
    return abi_encode(
        ["uint256", "uint256", "uint8", "bytes32", "bytes32"],
        [_to_int(value), _to_int(deadline), int(v), _to_bytes(r), _to_bytes(s)],
    )


def build_permit2_collector_data(
    nonce: str,
    deadline: str,
    permit2_signature: str,
    eip2612_permit_data: bytes | str = b"",
) -> bytes:
    """Encode `collectorData` for `Permit2DepositCollector.collect()`.

    `abi.encode(uint256 nonce, uint256 deadline, bytes permit2Sig, bytes eip2612Permit)`
    """
    if isinstance(eip2612_permit_data, str):
        eip2612 = _to_bytes(eip2612_permit_data) if eip2612_permit_data else b""
    else:
        eip2612 = eip2612_permit_data

    return abi_encode(
        ["uint256", "uint256", "bytes", "bytes"],
        [_to_int(nonce), _to_int(deadline), _to_bytes(permit2_signature), eip2612],
    )


__all__ = [
    "build_erc3009_deposit_nonce",
    "build_erc3009_collector_data",
    "build_eip2612_permit_data",
    "build_permit2_collector_data",
]

File: typescript/packages/mechanisms/evm/src/batch-settlement/encoding.ts
/**
 * @file Encoding helpers for batch-settlement deposit collectors.
 */
import { encodeAbiParameters, keccak256 } from "viem";

/**
 * Computes the ERC-3009 nonce used by the deposit collector:
 * `keccak256(abi.encode(channelId, salt))`.
 *
 * @param channelId - The `bytes32` channel id binding the authorization to a channel.
 * @param salt - Random salt provided by the client to make the nonce unique per deposit.
 * @returns The `bytes32` ERC-3009 nonce.
 */
export function buildErc3009DepositNonce(
  channelId: `0x${string}`,
  salt: `0x${string}`,
): `0x${string}` {
  return keccak256(
    encodeAbiParameters([{ type: "bytes32" }, { type: "uint256" }], [channelId, BigInt(salt)]),
  );
}

/**
 * Encodes the `collectorData` payload for `ERC3009DepositCollector.collect()`:
 * `abi.encode(validAfter, validBefore, salt, signature)`.
 *
 * @param validAfter - Earliest unix timestamp the authorization is valid (decimal string).
 * @param validBefore - Latest unix timestamp the authorization is valid (decimal string).
 * @param salt - Random salt provided by the client (hex string).
 * @param signature - ERC-3009 `ReceiveWithAuthorization` signature.
 * @returns ABI-encoded collector data passed to `deposit(..., collector, collectorData)`.
 */
export function buildErc3009CollectorData(
  validAfter: string,
  validBefore: string,
  salt: `0x${string}`,
  signature: `0x${string}`,
): `0x${string}` {
  return encodeAbiParameters(
    [{ type: "uint256" }, { type: "uint256" }, { type: "uint256" }, { type: "bytes" }],
    [BigInt(validAfter), BigInt(validBefore), BigInt(salt), signature],
  );
}

/**
 * Encodes optional EIP-2612 permit data consumed by `Permit2DepositCollector`.
 *
 * @param params - Permit amount, deadline, and split signature fields.
 * @param params.value - Approved Permit2 allowance value.
 * @param params.deadline - EIP-2612 permit deadline.
 * @param params.v - Signature recovery id.
 * @param params.r - Signature `r` value.
 * @param params.s - Signature `s` value.
 * @returns ABI-encoded permit segment.
 */
export function buildEip2612PermitData(params: {
  value: string;
  deadline: string;
  v: number;
  r: `0x${string}`;
  s: `0x${string}`;
}): `0x${string}` {
  return encodeAbiParameters(
    [
      { type: "uint256" },
      { type: "uint256" },
      { type: "uint8" },
      { type: "bytes32" },
      { type: "bytes32" },
    ],
    [BigInt(params.value), BigInt(params.deadline), params.v, params.r, params.s],
  );
}

/**
 * Encodes the `collectorData` payload for `Permit2DepositCollector.collect()`.
 *
 * @param nonce - Permit2 transfer nonce.
 * @param deadline - Permit2 transfer deadline.
 * @param permit2Signature - Signature over the channel-bound Permit2 authorization.
 * @param eip2612PermitData - Optional encoded EIP-2612 permit segment.
 * @returns ABI-encoded collector data passed to `deposit`.
 */
export function buildPermit2CollectorData(
  nonce: string,
  deadline: string,
  permit2Signature: `0x${string}`,
  eip2612PermitData: `0x${string}` = "0x",
): `0x${string}` {
  return encodeAbiParameters(
    [{ type: "uint256" }, { type: "uint256" }, { type: "bytes" }, { type: "bytes" }],
    [BigInt(nonce), BigInt(deadline), permit2Signature, eip2612PermitData],
  );
}
